vortex-cxx/src/lib.rs

```diff
@@ -0,0 +1,339 @@
+    let rt = get_runtime();
+    let array = rt
+        .block_on(async {
+            let stream = builder.inner.into_array_stream()?;
+            stream.read_all().await
+        })?
+        .into_arrow_preferred()?;
+
+    let ffi_array = FFI_ArrowArray::new(&array.to_data());
+    let ffi_schema = FFI_ArrowSchema::try_from(array.data_type())?;
+    // Two discarded attempts recorded here for future reference:
+    // 1. Require unsafe transmute and an intermediate CArrowArrayStream defined in cxx shared types
+    // Ok(unsafe { std::mem::transmute::<FFI_ArrowArrayStream, ffi::CArrowArrayStream>(stream) })
+    // 2. Require Box::new to heap allocate. Also requires an extern "Rust" type `ArrowArrayStream`.
+    // Ok(Box::new(ArrowArrayStream { inner: stream }))
+    let out_array = out_array as *mut FFI_ArrowArray;
+    let out_schema = out_schema as *mut FFI_ArrowSchema;
+    // # Safety
+    // Arrow C ABI
+    std::ptr::write(out_array, ffi_array);
+    std::ptr::write(out_schema, ffi_schema);
+    Ok(())
+}
+
+/// # Safety
+///
+/// out_stream should be properly aligned according to C ABI and valid for write.
+unsafe fn scan_builder_into_stream(
+    builder: Box<VortexScanBuilder>,
+    out_stream: *mut u8,
+) -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
+    let iter =
+        array_stream_to_iterator(builder.inner.into_array_stream()?.boxed() as SendableArrayStream);
+    let reader = VortexRecordBatchReader::try_new(iter)?;
+    let stream = FFI_ArrowArrayStream::new(Box::new(reader));
+    let out_stream = out_stream as *mut FFI_ArrowArrayStream;
+    // # Safety
+    // Arrow C ABI
+    std::ptr::write(out_stream, stream);
+    Ok(())
+}
+
+pub struct VortexWriteOptions {
+    inner: WriteOptions,
+}
+
+fn write_options_new() -> Box<VortexWriteOptions> {
+    Box::new(VortexWriteOptions {
+        inner: WriteOptions::default(),
+    })
+}
+
+/// # Safety
+///
+/// input_stream should be valid FFI_ArrowArrayStream.
+/// See [`FFI_ArrowArrayStream::from_raw`]
+unsafe fn write_array_stream(
+    options: Box<VortexWriteOptions>,
+    input_stream: *mut u8,
+    path: &str,
+) -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
+    let stream_reader =
+        ArrowArrayStreamReader::from_raw(input_stream as *mut FFI_ArrowArrayStream)?;
+
+    let rt = get_runtime();
+    let path = path.to_string();
+
+    // TODO: this actually materializes all the data into memory, should expose async write API to C++, see vortex-ffi/src/sink.rs
+    // Read all data from the Arrow stream synchronously before entering async context
+    let mut arrays: Vec<ArrayRef> = Vec::new();
+    let schema = stream_reader.schema();
+    let dtype = DType::from_arrow(schema);
+
+    // Convert Arrow RecordBatches to Vortex Arrays synchronously
+    for batch_result in stream_reader {
+        let batch = batch_result?;
+        let vortex_array = ArrayRef::from_arrow(batch, false);
+        arrays.push(vortex_array);
+    }
```

Comment:
This will buffer a lot of data in memory, can convert the stream  into `SendableArrayStream = Pin<Box<dyn ArrayStream + Send>>` and write that to the file.

Reply:
I tried converting the stream to SendableArrayStream before https://github.com/vortex-data/vortex/pull/3792/commits/8e3965f7ab957be60a7bd73f501177781b6eab21#diff-7185d60946e047abb9c984a9f8984aa4cb37983ac3d5d2528ef03909c3dfba49R158-R169

But it has the same problem as https://github.com/vortex-data/vortex/issues/3877

Reply:
We can at least use a model (writing via a channel) similar to hear: https://github.com/vortex-data/vortex/blob/6271c6e54371c650183a9d4f61f34a000651a863/vortex-ffi/src/sink.rs#L71 which means that the whole array is not buffered in memory

Reply:
@gatesn shall we wait for the threading model to be cleaned up before merging this PR?

Reply:
I see. Thanks! Will check it out.

Reply:
Ive fixed it. Would you mind take a look again?